// pinus-loader/dist/lib/decoraters.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isDefined = exports.isCronner = exports.isHandler = exports.isRemoter = exports.method = exports.notImplement = exports.cronner = exports.handler = exports.remoter = exports.DESIGN_PARAM_TYPES = exports.LoaderPathType = void 0;
var LoaderPathType;
(function (LoaderPathType) {
    LoaderPathType["PINUS_REMOTER"] = "pinus:remoter";
    LoaderPathType["PINUS_HANDLER"] = "pinus:hanlder";
    LoaderPathType["PINUS_CRONNER"] = "pinus:cronner";
})(LoaderPathType = exports.LoaderPathType || (exports.LoaderPathType = {}));
const DUPLICATED_REMOTER_DECORATOR = 'Cannot apply @remoter decorator multiple times.';
const DUPLICATED_HANDLER_DECORATOR = 'Cannot apply @handler decorator multiple times.';
const DUPLICATED_CRONNER_DECORATOR = 'Cannot apply @cronner decorator multiple times.';
// used to access design time types
exports.DESIGN_PARAM_TYPES = 'design:paramtypes';
function remoter() {
    return function (target) {
        if (Reflect.hasOwnMetadata(LoaderPathType.PINUS_REMOTER, target)) {
            throw new Error(DUPLICATED_REMOTER_DECORATOR);
        }
        const types = Reflect.getMetadata(exports.DESIGN_PARAM_TYPES, target) || [];
        Reflect.defineMetadata(LoaderPathType.PINUS_REMOTER, types, target);
        return target;
    };
}
exports.remoter = remoter;
function handler() {
    return function (target) {
        if (Reflect.hasOwnMetadata(LoaderPathType.PINUS_HANDLER, target)) {
            throw new Error(DUPLICATED_HANDLER_DECORATOR);
        }
        const types = Reflect.getMetadata(exports.DESIGN_PARAM_TYPES, target) || [];
        Reflect.defineMetadata(LoaderPathType.PINUS_HANDLER, types, target);
        return target;
    };
}
exports.handler = handler;
function cronner() {
    return function (target) {
        if (Reflect.hasOwnMetadata(LoaderPathType.PINUS_CRONNER, target)) {
            throw new Error(DUPLICATED_CRONNER_DECORATOR);
        }
        const types = Reflect.getMetadata(exports.DESIGN_PARAM_TYPES, target) || [];
        Reflect.defineMetadata(LoaderPathType.PINUS_CRONNER, types, target);
        return target;
    };
}
exports.cronner = cronner;
function notImplement() {
    throw new Error('not implement');
}
exports.notImplement = notImplement;
function method() {
    return function (target, targetKey, index) {
        return {
            configurable: true,
            enumerable: true,
            value: notImplement,
            writable: true
        };
    };
}
exports.method = method;
function isRemoter(ctor) {
    return Reflect.hasMetadata(LoaderPathType.PINUS_REMOTER, ctor);
}
exports.isRemoter = isRemoter;
function isHandler(ctor) {
    return Reflect.hasMetadata(LoaderPathType.PINUS_HANDLER, ctor);
}
exports.isHandler = isHandler;
function isCronner(ctor) {
    return Reflect.hasMetadata(LoaderPathType.PINUS_CRONNER, ctor);
}
exports.isCronner = isCronner;
function isDefined(ctor, pathType) {
    return Reflect.hasMetadata(pathType, ctor);
}
exports.isDefined = isDefined;
