// pinus-loader@1.7.1/dist/lib/container.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.removeFromContainer = exports.getFromContainer = exports.isUseContainer = exports.useContainer = exports.defaultContainer = void 0;
/**
 * Container to be used by this library for inversion control. If container was not implicitly set then by default
 * container simply creates a new instance of the given class.
 */
exports.defaultContainer = new (class {
    constructor() {
        this.instances = [];
    }
    get(someClass) {
        let instance = this.instances.find(instance => instance.type === someClass);
        if (!instance) {
            instance = { type: someClass, object: new someClass() };
            this.instances.push(instance);
        }
        return instance.object;
    }
    unbind(someClass) {
        let idx = this.instances.findIndex(instance => instance.type === someClass);
        if (idx >= 0) {
            this.instances.splice(idx, 1);
        }
    }
})();
let userContainer;
let userContainerOptions;
/**
 * Sets container to be used by this library.
 */
function useContainer(iocContainer, options) {
    userContainer = iocContainer;
    userContainerOptions = options;
}
exports.useContainer = useContainer;
function isUseContainer() {
    return userContainer !== undefined;
}
exports.isUseContainer = isUseContainer;
/**
 * Gets the IOC container used by this library.
 */
function getFromContainer(someClass) {
    if (userContainer) {
        try {
            const instance = userContainer.get(someClass);
            if (instance)
                return instance;
            if (!userContainerOptions || !userContainerOptions.fallback)
                return instance;
        }
        catch (error) {
            if (!userContainerOptions || !userContainerOptions.fallbackOnErrors)
                throw error;
        }
    }
    return exports.defaultContainer.get(someClass);
}
exports.getFromContainer = getFromContainer;
/**
 * Remove object from the IOC container used by this library.
 */
function removeFromContainer(someClass) {
    if (userContainer) {
        try {
            userContainer.unbind(someClass);
        }
        catch (error) {
            if (!userContainerOptions || !userContainerOptions.fallbackOnErrors)
                throw error;
        }
    }
    exports.defaultContainer.unbind(someClass);
}
exports.removeFromContainer = removeFromContainer;
